litespi/gen.py

```diff
@@ -0,0 +1,181 @@
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+LiteSPI standalone core generator
+
+LiteSPI aims to be directly used as a python package when the SoC is created using LiteX. However,
+for some use cases it could be interesting to generate a standalone verilog file of the core:
+- integration of the core in a SoC using a more traditional flow.
+- need to version/package the core.
+- avoid Migen/LiteX dependencies.
+- etc...
+
+The standalone core is generated from a YAML configuration file that allows the user to generate
+easily a custom configuration of the core.
+
+"""
+
+import argparse
+import os
+import yaml
+
+from migen import *
+
+from litex.build.generic_platform import *
+from litex.build.xilinx.platform import XilinxPlatform
+
+from litex.soc.interconnect import wishbone
+from litex.soc.interconnect import axi
+from litex.soc.integration.soc_core import *
+from litex.soc.integration.builder import *
+from litex.soc.integration.soc import SoCRegion
+
+import inspect
+import litespi.modules
+from litespi.spi_nor_flash_module import SpiNorFlashModule
+
+# module database
+modules_dict = {name: cls for name, cls in inspect.getmembers(litespi.modules) if inspect.isclass(cls) and issubclass(cls, SpiNorFlashModule)}
+
+
+# IOs ----------------------------------------------------------------------------------------------
+
+_io = [
+    # Clk / Rst
+    ("sys_clock", 0, Pins(1)),
+    ("sys_reset", 1, Pins(1)),
+
+    # SPI-Flash (X1).
+    ("spiflash", 0,
+        Subsignal("cs_n", Pins(1)),
+        Subsignal("clk",  Pins(1)),
+        Subsignal("mosi", Pins(1)),
+        Subsignal("miso", Pins(1)),
+        Subsignal("wp",   Pins(1)),
+        Subsignal("hold", Pins(1)),
+    ),
+
+    # SPI-Flash (X4).
+    ("spiflash4x", 0,
+        Subsignal("cs_n", Pins(1)),
+        Subsignal("clk",  Pins(1)),
+        Subsignal("dq",   Pins(4)),
+    ),
+]
+
+class SPICore(SoCMini):
+    def __init__(self, platform, core_config):
+
+        # SoC parameters ---------------------------------------------------------------------------
+        soc_args = {}
+        if "soc" in core_config:
+            soc_config = core_config["soc"]
+
+            for arg in soc_config:
+                if arg in ("csr_map", "interrupt_map", "mem_map"):
+                    getattr(self, arg).update(soc_config[arg])
+                else:
+                    soc_args[arg] = soc_config[arg]
+
+        # SoCMini ----------------------------------------------------------------------------------
+        SoCMini.__init__(self, platform, clk_freq=int(1e6), **soc_args)
+
+        # CRG --------------------------------------------------------------------------------------
+        self.submodules.crg = CRG(platform.request("sys_clock"), platform.request("sys_reset"))
+
+        device = core_config["device"]
+        mode = core_config["mode"]
+        assert mode[-1] == "x" #expected parameter should end in "x" (typically 1x or 4x)
```

Comment:
Please prefer something like this:` assert mode in ["x1", "x4]`, this is more direct, easier to read and avoid comments.